src/librustc_data_structures/dependency/mod.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Dependency graph used for incremental compilation.
+
+use std::cell::RefCell;
+use std::fmt::Debug;
+use std::hash::Hash;
+use self::state::DepGraphState;
+
+mod state;
+
+#[cfg(test)]
+mod test;
+
+pub struct DepGraph<ID: DepNodeId> {
```

Comment:
@michaelwoerister 

> Do I understand this correctly, that this parameterization over DepNodeId only exists because DefId is defined in a downstream crate?

Yes, basically. But I also think it's nice to have the `DepGraph` as a standalone data structure that doesn't drag in the particular details of the compiler too much.